[3811] Add findable to with_vacancies scope for courses

SamJamCul opened the pull request:
### Context
https://trello.com/c/ejcChJPq/3811-change-with-has-vacancies-to-only-include-findable-courses

### Changes proposed in this pull request
* Add findable scoping to the with_vacancies scope for courses

### Guidance to review

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

SamJamCul commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Change with_/has_ vacancies to only include findable courses](https://trello.com/c/ejcChJPq/3811-change-with-has-vacancies-to-only-include-findable-courses)

SamJamCul commented:
> Looks good. Can we also change `SiteStatus#with_vacancies` to keep things consistent, though?

Done, also changed the SiteStatus `with_vacancies?` as well

defong commented on app/models/course.rb:
```diff
@@ -183,7 +183,7 @@ def new_draft_attributes
   scope :with_recruitment_cycle, ->(year) { joins(provider: :recruitment_cycle).where(recruitment_cycle: { year: year }) }
   scope :findable, -> { joins(:site_statuses).merge(SiteStatus.findable) }
-  scope :with_vacancies, -> { joins(:site_statuses).merge(SiteStatus.with_vacancies) }
+  scope :with_vacancies, -> { joins(:site_statuses).merge(SiteStatus.findable.with_vacancies) }
```
was this meant to be changed?

given 

```ruby
# app/models/site_status.rb 
scope :with_vacancies, -> { where.not(vac_status: :no_vacancies).findable }
```

SamJamCul replied:
You're right, it shouldn't need to be changed seeing as site status already covers it

misaka commented:
Just noticed https://github.com/DFE-Digital/teacher-training-api/blob/3811-change-with-has-vacancies-to-only-include-findable-course/app/models/course.rb#L498-L502

```ruby
  def ensure_site_statuses_match_study_mode
    site_statuses.select(&:with_vacancies?).each do |site_status|
      update_vac_status(study_mode, site_status)
    end
  end
```

We've changed how `with_vacancies?` behaves, have we thought through whether that will be an issue?

misaka commented:
Is it worth changing this bit too? It sounds like having `findable` applied twice will be a null-op, but it could also be confusing to the reader.

https://github.com/DFE-Digital/teacher-training-api/blob/master/app/services/publish_reporting_service.rb#L109-L110

```ruby
    findable_courses = courses_changed_recently.findable.distinct
    open_courses = findable_courses.with_vacancies
```

SamJamCul commented:
> Is it worth changing this bit too? It sounds like having `findable` applied twice will be a null-op, but it could also be confusing to the reader.
> 
> https://github.com/DFE-Digital/teacher-training-api/blob/master/app/services/publish_reporting_service.rb#L109-L110
> 
> ```ruby
>     findable_courses = courses_changed_recently.findable.distinct
>     open_courses = findable_courses.with_vacancies
> ```

For this one we did spot it, but decided against changing it, as the `findable_courses` bit is used elsewhere, as well as having the `distinct` on the end. It felt like it would be more hassle adding `findable` to the other uses rather than just having this redundant one here.

SamJamCul commented:
> Just noticed https://github.com/DFE-Digital/teacher-training-api/blob/master/app/models/course.rb#L498-L502
> 
> ```ruby
>   def ensure_site_statuses_match_study_mode
>     site_statuses.select(&:with_vacancies?).each do |site_status|
>       update_vac_status(study_mode, site_status)
>     end
>   end
> ```
> 
> We've changed how `with_vacancies?` behaves, have we thought through whether that will be an issue?

Fixed this, we're now checking it slightly differently, by looking directly at each site statuses vac_status. This bypasses the `with_vacancies?` findable problem, keeping the original functionality intact!

misaka reviewed (approved):
Good work guys, this was tricky, but is one step closer to Perfect Datas!!! 💯
